packages/@glimmer/compiler/lib/template-compiler.ts

```diff
@@ -62,6 +66,7 @@ export default class TemplateCompiler implements Processor<InputOps> {
   }
 
   startProgram([program]: [AST.Template]) {
+    this.cursorCount = 0;
```

Comment:
~~I don't fully understand why we are resetting this counter for each program. I would have assumed that it was important that the `guid` we used was unique across each template (or is it needing to be more globally unique?), but resetting in `startProgram` here will lead to the same template reusing the same cursorCount (and therefore `this.cursor()` not being unique across the template).~~

D'oh! I just realized (after digging more into the code) that `startProgram` here is super misleading. It _really_ should have been renamed to `startTemplate` when we split the usages of `Program` AST nodes into `Template` and `Block`.

I do still wonder what level of uniqueness we are looking for with the guid, but that is somewhat unrelated.

Reply:
Somewhat ironically, the _original_ code did suffer from the problem I described since it is operating on the `Handlebars.Program` (which is definitely emitted for each branch of a block). 

See [this AST Explorer](https://astexplorer.net/#/gist/16cff7a9560a43d914d289168c988ed1/) showing that multiple `Program` nodes would have been seen in the `HandlebarsNodeVisitors` and therefore the same cursor would be reused within a template like 

```hbs
{{#if foo}}
  {{#in-element this.el}}{{/in-element}}
{{else}}
  {{#in-element this.el}}{{/in-element}}
{{/if}}
```